### ISSUE TITLE
Schema module with mandatory choice cannot be installed

### ISSUE THREAD

**Issue by pfeige** [CESNET/libyang#1804]
Hi,

when trying to install the following schema module:
~~~
module issue-mandatory-choice {
  namespace "urn:issue:mandatory-choice";
  prefix imc;

  container ca {
    container sam {
      leaf publickey {
        type empty;
      }
      leaf password {
        type empty;
      }
      leaf none {
        type empty;
      }
    }
    choice loe {
      mandatory true;
      case loc {
        container users {
          list user {
            key name;
            leaf name {
              type string;
            }
            leaf password {
              type string;
            }
          }
        }
      }
      case ext {
        leaf cade {
          type empty;
        }
      }
    }
  }
}
~~~
the following error is shown:
~~~
[ERR] Mandatory choice "loe" data do not exist.
~~~
There are no data for this module available yet.

BR,
Peter

**Comment by michalvasko** [comment]
> There are no data for this module available yet.

Yes, but there need to be. Look at `-I` parameter of `sysrepoctl`.

**Comment by pfeige** [comment]
Is it really necessary to provide data when just installing modules? Without that mandatory statement the module is installed without problems.
And when I execute "sysrepoctl -i /yang/issue-madatory-choice.yang  -I /yang/issue-mandatory-choice.xml" with the following data:
~~~
<ca xmlns="urn:issue:mandatory-choice">
  <sam>
    <password/>
    <publickey/>
  </sam>
  <users/>
</ca>
~~~
the same error occurs.

**Comment by michalvasko** [comment]
> Is it really necessary to provide data when just installing modules?

Yes, it really is. Otherwise, you would make all the datastores invalid, how is *sysrepo* supposed to handle that if it guarantees validity? It is true that *sysrepo* v0.7.x supported this use-case and hacked around it but I saw no reason for doing that. Also, it is the "fault" of the YANG module that it is not valid with no data. Either it is intentional when requiring the initial data is desired (intentional, again) or it is not when it is the problem of the module.

> the following data: ... the same error occurs.

This will probably be due to the container `users` not having `presence` (it [should](https://github.com/netmod-wg/yang-next/issues/88), ideally).

**Comment by pfeige** [comment]
You mean the empty container users does not fulfill the mandatory statement of the choice if it has no presence statement?

**Comment by michalvasko** [comment]
Yes, I suppose that is what I am saying, details are in the mailing list discussions referenced in the issue I referenced. But I understand you do not feel like going through all that so I can explain with the YANG [specs](https://datatracker.ietf.org/doc/html/rfc7950#section-7.5.1), second paragraph.

**Comment by pfeige** [comment]
OK, I see.
